### DISCUSSION TITLE
How to insert youtube iframe on pasting youtube links

### DISCUSSION THREAD

**Discussion by fiedl** [ueberdosis/tiptap#816] [General Questions]
When the user pastes a youtube link like https://www.youtube.com/watch?v=..., this video should be embedded as iframe as shown in https://tiptap.dev/embeds.

<img width="969" alt="Bildschirmfoto 2020-08-31 um 11 02 54" src="https://user-images.githubusercontent.com/1679688/91702791-88634a00-eb79-11ea-8e06-232b48ae0ae9.png">

Does anyone have a working solution for that?

Thank your!

### Work in progress

https://github.com/fiedl/your_platform/commit/0a603b03f51c9f00e677a6831912c9742874ef7c

Does not work yet:

- [ ] Responsive iframe — because I didn't find a way to *save* the wrapper div to the html value
- [ ] If the iframe is the only content, I can't select it, e.g. in order to delete it again

### Resources

- https://github.com/ueberdosis/tiptap/blob/main/examples/Components/Routes/Embeds/Iframe.js
- https://github.com/ueberdosis/tiptap/issues/88
- https://github.com/ueberdosis/tiptap/issues/112
- https://github.com/ueberdosis/tiptap/issues/689

**Comment by softmarshmallow** [comment]
any good news?

**Comment by LazAustin** [comment]
Any solutions on making this extension responsive?

**Comment by mapszk** [comment]
I used the oembed API from youtube that returns the html ready to embed from a youtube link, then inserted it with insertContent()

**Reply by piotrkulpinski** [reply]
Could you share the code you've used to insert the HTML content? I'm trying to do the same but can't get it to insert div elements.

**Reply by mapszk** [reply]
With the video url that the user submits I make a request to the oembed API that returns the HTML
```
export const getOEmbedHtml = async (oembedUrl) => {
  // request the HTML code to embed 
  const response = await axios.get(oembedUrl).catch(e => e.response)
  const successfulRequest = !isError(response.status)
  if (successfulRequest) {
    // convert it to HTML
    const elem = new DOMParser().parseFromString(response.data.html, 'text/html')
    // look for the iframe to resize it instead of using the fixed width and height that returns youtube
    const iframe = elem.querySelector('iframe')
    iframe.setAttribute('height', '100%')
    iframe.setAttribute('width', '100%')
    return iframe.outerHTML
  } else return null
}
```
That function would only works with websites that return an iframe.

Then in my text editor menu I insert the content:

```
const html = await getOEmbedHtml(oembedUrl)
if (html) {
  this.editor.commands.insertContent(html, {
    parseOptions: {
      preserveWhitespace: false
    }
  })
}
```

Hope it helps

**Reply by piotrkulpinski** [reply]
Thanks a lot!
Unfortunately, I still can't get it to insert my iframe html into the editor. 
What version of TipTap do you have? Is there any special config option I need to set?

**Reply by mapszk** [reply]
I'm using this [iframe extension](https://tiptap.dev/experiments/embeds), just copied the code and adapted to javascript

**Comment by bookercodes** [comment]
You can use the [YouTube extension](https://tiptap.dev/docs/editor/extensions/nodes/youtube) for this purpose. It has a `addPasteHandler` option (`true` by default)

Here's the implementation code in case the YouTube extension isn't suitable for you but you'd like to learn how it's done:


```
addPasteRules() {
    if (!this.options.addPasteHandler) {
      return []
    }

    return [
      nodePasteRule({
        find: YOUTUBE_REGEX_GLOBAL,
        type: this.type,
        getAttributes: match => {
          return { src: match.input }
        },
      }),
    ]
  },
```

**Reply by AbdulWahhab66** [reply]
This really helped. I was having an issue where pasting youtube homepage link was also creating an embed. Adding a nodepasterule to ignore it did the job for me.
